### ISSUE TITLE
issue with RS density fit for periodic systems

### ISSUE THREAD

**Issue by lkwagner** [pyscf/pyscf#1335]
Hi all,

I was testing the range-separated density fits on calcium oxychloride, copied here. I found that SCF, while it converges for GDF, does not converge at all for RS-DF, which was surprising to me. If we run with a kmesh of 111, they both run, converge the same,  and agree on energy. I could not reproduce this on a simpler example system unfortunately. 

[test_rs.tar.gz](https://github.com/pyscf/pyscf/files/9160133/test_rs.tar.gz)

**Comment by tberkel** [comment]
Thanks, we'll take a look. Sometimes the improved precision of RSDF actually amplifies numerical issues, e.g. linear dependencies in the basis set or the aux basis. In these cases, the success of GDF is somewhat fortuitous. Not sure if that's what's going on here though.

**Comment by hongzhouye** [comment]
Hi Lucas, the issue is with the AFTDF code in PySCF which is independent of RSDF. Particularly, if you replace this [line](https://github.com/pyscf/pyscf/blob/master/pyscf/pbc/df/rsdf.py#L364) with
```Python
bvk_kmesh=None
```
RSDF converges to the same answer as GDF (For the SCF energy, I got `-1314.89735407836` with RSDF which matches your GDF number `-1314.89735374591` to about 1e-7 Ha).

I haven't managed to pinpoint exactly which part of the AFTDF code caused this issue (@sunqm any thoughts?). But for now, changing the line I mentioned above is enough to get RSDF to run.

**Comment by hongzhouye** [comment]
@sunqm Particularly, the issue is with the BvK code in the AFT module. I found in quite a few systems that when **the unit cell is large**, the AO-pair AFTs are not accurate when using BvK. The system Lucas used is a good example: even a k-point mesh as small as 2x2x1 shows this problem.

**Comment by lkwagner** [comment]
Thanks for looking into this so quickly! I'll be excited to use it when it's resolved.

**Comment by hongzhouye** [comment]
@lkwagner will let you know when the bug with AFT is resolved!
But as I said, there's no bug related to RSDF at this point. So if you don't minding hacking a tiny bit, the one line change I mentioned earlier is sufficient for now.

**Comment by gkc1000** [comment]
Maybe sth to do with _estimate_overlap? 
in
https://github.com/pyscf/pyscf/blob/290413d451c3b5b10d16a1d49b90285ef2c9bfd7/pyscf/pbc/df/ft_ao.py#L188

**Comment by sunqm** [comment]
@lkwagner Which branch do you use? I tried this system with branch https://github.com/pyscf/pyscf/tree/rsjk-rsdf. Your script gives -1314.89737363138. Although slightly different to Hongzhou's results, it does not get any convergence problem.

I'm not particularly sure what is the main cause of the accuracy problem. bvk_mesh and overlap estimation might be two sources. However, relevant estimations are treated very conservative. Each quantity itself should be converged tightly to the precision required (or at least very closed to the required precision).  I'd suspect the discrepancy is caused by the accuracy of the 2-center Coulomb integrals. Because it's basically the denominator in density fitting errors may be enlarged. In this regards, tuning the density fitting auxiliary basis may be able to change the accuracy. Many possibilities have to be tested.

For the moment a quick workaround is to tighten the accuracy parameter `cell.precision=1e-12` . It will slow down the RSDF construction but it should still be faster than the GDF implementations on the main branch.

**Comment by lkwagner** [comment]
@sunqm , I used the main branch. To be clear, we are fine just using GDF for the moment; we don't need a workaround. We just found behavior that we thought you'd like to address and so reported it.

**Comment by sunqm** [comment]
This problem is caused by the numerical issues in the eigenvalue solver and other matrix decomposition procedure for complex matrix. Even every quantity is converged to the desired precision, I can still observe non-negligible errors in the 3-index tensor cderi. Error becomes especially big when linear dependency is found in the auxiliary basis because the overlap of aux basis has contribution to the denominator in DF tensor decomposition. GDF does not cure the issue. It just happens to be stable in some tests due to the GDF implementation in v2.0. In this version quantities are converged far more tightly than the required precision and this leads to relatively small errors (and this is also the reason GDF is slow in v2.0). When "better" integral screening was used similar numerical problems are also observed in GDF. It seems quite difficult to overcome the issue simply by the implementation or the algorithm. Aux basis suitable for periodic systems needs to be developed.

**Comment by hongzhouye** [comment]
Hi @sunqm were you talking about specifically Lucas' example? I asked bc in his example script, he used `cell.exp_to_discard = 0.2`, which is very conservative and essentially removes any linear dependency issue in the basis set (for example, the AO basis overlap matrix has a condition number of only 300, while the usual "problematic" case I found has a condition number close to 1e10) and the corresponding fitting ETB.

**Comment by sunqm** [comment]
Yes. I'm talking about the error in the 2-center 2-electron integrals for the auxiliary basis. Errors are more significant than other treatments. Even each quantity (the 2-center integral matrix, the 3-center integral tensor) converges to the required accuracy, the generated 3-index tensor has larger errors which causes uncertainty in SCF results at the more significant decimal place.

**Comment by ElonDormancy** [comment]
This issue is caused by the **aft module bug of pyscf v2.0.0** but in version 2.1.0 the bug has fixed. So the fastest way to solve this problem is to use pyscf of the lastest version(>=2.1.0) or you can set `bvk_kmesh=None` to use the non bvk algorithm as hongzhou suggested. If you insist on using the bvk_kmesh version of pyscf v2.0.0 ,you can fix the bug as following:
File:pyscf/lib/pbc/ft_ao.c(pyscf v2.0.0)
Two function have the same issue(`_ft_bvk_k` and `_ft_bvk_nk1`),the offset of `ovlp_mask` is wrong.
```
/*
 * Multiple k-points for BvK cell
 */
static void _ft_bvk_k(int (*intor)(), int (*eval_aopair)(), void (*eval_gz)(),
                      void (*fsort)(), double complex *out,
                      int nkpts, int comp, int nimgs, int bvk_nimgs, int blksize,
                      int ish, int jsh, int *cell_loc_bvk, char *ovlp_mask,
                      double complex *buf, double *env_loc, double *Ls,
                      double complex *expkL, int *shls_slice, int *ao_loc,
                      double *sGv, double *b, int *sgxyz, int *gs, int nGv,
                      int *atm, int natm, int *bas, int nbas, double *env)
{
        const int ish0 = shls_slice[0];
        const int jsh0 = shls_slice[2];
        const int jsh1 = shls_slice[3];
        const int njsh = jsh1 - jsh0;
        ovlp_mask += (ish * njsh + jsh) * nimgs;
...
}
```
The shape of ovlp_mask is (nbas,nbas,nimgs),and shls_slice[:4] = (ish0,ish1,jsh0,jsh1)[ish0,ish,jsh0,jsh1 \in (0,nbas)] means that we calculate the slice (ish0-ish1,jsh0-jsh1).So the code above should be corrected:
```
{
        const int ish0 = shls_slice[0];
        const int ish1 = shls_slice[1];
        const int jsh0 = shls_slice[2];
        const int jsh1 = shls_slice[3];
        int offset_ish = ish0 + ish;
        int offset_jsh = jsh0 - nbas + jsh;
        ovlp_mask += (offset_ish * nbas + offset_jsh) * nimgs;
...
}
```
as the same the `_ft_bvk_nk1` function also needs to be corrected.

Before correction:converged SCF energy = -1378.44207434398(wrong result)
After correction:converged SCF energy = -1314.89735407836, which matches hongzhou's result.

**Comment by sunqm** [comment]
@ElonDormancy thank you for analyzing the problem and finding the bug! Could you please create a PR to fix the bug?

**Comment by verena-neufeld** [comment]
Hi @lkwagner , thanks for raising this issue. @hongzhouye @sunqm @ElonDormancy has it been fixed now with PR #1622 ?

**Comment by sunqm** [comment]
Feel free to reopen the issue if errors persist in pyscf-2.6 and newer.
